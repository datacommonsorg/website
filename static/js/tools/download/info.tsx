/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Info page before a chart is shown.
 */

import React from "react";

// Data object used to customize the example rendered by this component.
// A single example place to display.
export interface InfoPlace {
  name: string;
  dcid: string;
}

// Data object used to customize the example rendered by this component.
interface InfoPropType {
  // NOTE: Exactly 2 InfoPlace objects are expected.
  infoPlaces: [InfoPlace, InfoPlace];
}

export function Info(props: InfoPropType): JSX.Element {
  return (
    <div id="placeholder-container">
      <p>
        The data download tool allows you to download data for the statistical
        variables in the left pane. To get started:
      </p>
      <ol>
        <li>
          Enter a place in the search box and then select the type of places you
          want to download data for.
        </li>
        <li>
          Pick up to 5 statistical variables in the left pane. There are
          thousands of statistical variables to choose from, arranged in a
          topical hierarchy.
        </li>
        <li>
          Choose whether you want the data for the latest date, all available
          dates, or a custom date range.
        </li>
      </ol>
      <p>
        Data will be downloaded as a .csv file where each row contains the data
        for a place and date combination. For example:
      </p>
      <div className="download-info-table-example">
        <table>
          <thead>
            <tr>
              <th>placeDcid</th>
              <th>placeName</th>
              <th>Date:[Variable1]</th>
              <th>Value:[Variable1]</th>
              <th>Source:[Variable1]</th>
              <th>Date:[Variable2]</th>
              <th>Value:[Variable2]</th>
              <th>Source:[Variable2]</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>{props.infoPlaces[0].dcid}</td>
              <td>{props.infoPlaces[0].name}</td>
              <td>2020-01</td>
              <td>1</td>
              <td>abc.com</td>
              <td>2020</td>
              <td>11</td>
              <td>xyz.com</td>
            </tr>
            <tr>
              <td>{props.infoPlaces[0].dcid}</td>
              <td>{props.infoPlaces[0].name}</td>
              <td>2020-02</td>
              <td>1</td>
              <td>abc.com</td>
              <td></td>
              <td></td>
              <td></td>
            </tr>
            <tr>
              <td>{props.infoPlaces[0].dcid}</td>
              <td>{props.infoPlaces[0].name}</td>
              <td>2020-03</td>
              <td>1</td>
              <td>abc.com</td>
              <td></td>
              <td></td>
              <td></td>
            </tr>
            <tr>
              <td>...</td>
            </tr>
            <tr>
              <td>{props.infoPlaces[0].dcid}</td>
              <td>{props.infoPlaces[0].name}</td>
              <td>2021-01</td>
              <td>1</td>
              <td>abc.com</td>
              <td>2021</td>
              <td>11</td>
              <td>xyz.com</td>
            </tr>
            <tr>
              <td>{props.infoPlaces[1].name}</td>
              <td>{props.infoPlaces[1].name}</td>
              <td>2020-01</td>
              <td>2</td>
              <td>abc.com</td>
              <td>2020</td>
              <td>22</td>
              <td>xyz.com</td>
            </tr>
            <tr>
              <td>{props.infoPlaces[1].name}</td>
              <td>{props.infoPlaces[1].name}</td>
              <td>2020-02</td>
              <td>2</td>
              <td>abc.com</td>
              <td></td>
              <td></td>
              <td></td>
            </tr>
            <tr>
              <td>...</td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  );
}
