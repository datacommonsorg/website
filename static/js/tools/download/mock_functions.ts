/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* mocked axios calls for Page test for download tool. */

jest.mock("axios");
import axios from "axios";
import { when } from "jest-when";

import { stringifyFn } from "../../utils/axios";

export function axios_mock(): void {
  // Mock all the async axios call.
  axios.get = jest.fn();
  axios.post = jest.fn();

  const rootGroupsData = {
    data: {
      childStatVarGroups: [
        {
          displayName: "Demographics",
          id: "dc/g/Demographics",
          specializedEntity: "Demographics",
          descendentStatVarCount: 2,
        },
        {
          displayName: "Economics",
          id: "dc/g/Economics",
          specializedEntity: "Economics",
          descendentStatVarCount: 2,
        },
      ],
    },
  };

  const demographicsGroupsData = {
    data: {
      childStatVarGroups: [],
      childStatVars: [
        {
          displayName: "Population",
          id: "Count_Person",
          searchName: "Count Of Person",
          hasData: true,
        },
        {
          displayName: "Median Age",
          id: "Median_Age_Person",
          searchName: "Median Age",
          hasData: true,
        },
      ],
    },
  };

  const csvData =
    "placeDcid,placeName,Date:Count_Person,Value:Count_Person,Source:Count_Person\ngeoId/06001,Alameda County,2021,1648556,https://www2.census.go\ngeoId/06002,Alpine County,2021,1235,https://www2.census.gov/";

  // get statvar properties Count_Person
  when(axios.get)
    .calledWith("/api/stats/stats-var-property?dcid=Count_Person")
    .mockResolvedValue({
      data: {
        Count_Person: {
          md: "",
          mprop: "count",
          pt: "Person",
          pvs: {},
          title: "Population",
        },
      },
    });

  // get statvar properties Median_Age_Person
  when(axios.get)
    .calledWith("/api/stats/stats-var-property?dcid=Median_Age_Person")
    .mockResolvedValue({
      data: {
        Median_Age_Person: {
          md: "",
          mprop: "",
          pt: "Person",
          pvs: {},
          title: "Median Age",
        },
      },
    });

  // get place types, geoId/06
  when(axios.get)
    .calledWith("/api/place/type/geoId/06")
    .mockResolvedValue({ data: "State" });

  // get place names, geoId/06
  when(axios.get)
    .calledWith("/api/place/name?dcids=geoId/06")
    .mockResolvedValue({ data: { "geoId/06": "California" } });

  // get root stat var group
  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Root",
      entities: [],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(rootGroupsData);

  // get root stat var group for places in geoId/06
  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Root",
      entities: ["geoId/06001", "geoId/06002"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(rootGroupsData);

  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Root",
      entities: ["geoId/06002", "geoId/06001"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(rootGroupsData);

  // get demographics stat var group for places in geoId/06
  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Demographics",
      entities: ["geoId/06001", "geoId/06002"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(demographicsGroupsData);

  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Demographics",
      entities: ["geoId/06002", "geoId/06001"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(demographicsGroupsData);

  // stat var info for demographics node
  when(axios.get)
    .calledWith("/api/variable/info", {
      params: {
        dcids: ["Count_Person", "Median_Age_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          placeTypeSummary: {
            type1: {
              numPlaces: 0,
              topPlaces: [],
            },
          },
        },
        Median_Age_Person: {
          placeTypeSummary: {
            type1: {
              numPlaces: 0,
              topPlaces: [],
            },
          },
        },
      },
    });

  // get stat var path for Count_Person
  when(axios.get)
    .calledWith("/api/variable/path?dcid=Count_Person")
    .mockResolvedValue({
      data: ["Count_Person", "dc/g/Demographics"],
    });

  // get stat var path for Median_Age_Person
  when(axios.get)
    .calledWith("/api/variable/path?dcid=Median_Age_Person")
    .mockResolvedValue({
      data: ["Median_Age_Person", "dc/g/Demographics"],
    });

  // get places in for counties in geoId/06
  when(axios.get)
    .calledWith("/api/place/places-in?dcid=geoId/06&placeType=County")
    .mockResolvedValue({
      data: {
        "geoId/06": ["geoId/06001", "geoId/06002"],
      },
    });

  // get parent place for geoId/06
  when(axios.get)
    .calledWith("/api/place/parent/geoId/06")
    .mockResolvedValue({
      data: ["country/USA"],
    });

  // get facets within place for Count_Person
  when(axios.get)
    .calledWith("/api/facets/within", {
      params: {
        childType: "County",
        maxDate: "latest",
        minDate: "latest",
        parentPlace: "geoId/06",
        statVars: ["Count_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          importName: "Facet1",
        },
      },
    });

  when(axios.get)
    .calledWith("/api/facets/within", {
      params: {
        childType: "County",
        maxDate: "",
        minDate: "2020",
        parentPlace: "geoId/06",
        statVars: ["Count_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          importName: "Facet1",
        },
      },
    });

  // get facets within place for Count_Person and Median_Age_Person
  when(axios.get)
    .calledWith("/api/facets/within", {
      params: {
        childType: "County",
        maxDate: "",
        minDate: "2020",
        parentPlace: "geoId/06",
        statVars: ["Count_Person", "Median_Age_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          importName: "Facet1",
        },
        Median_Age_Person: {
          importName: "Facet1",
        },
      },
    });

  // get place stats vars for places in geoId/06
  when(axios.post)
    .calledWith("/api/place/stat-vars/union", {
      dcids: ["geoId/06001", "geoId/06002"],
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: ["Count_Person"],
    });

  when(axios.post)
    .calledWith("/api/place/stat-vars/union", {
      dcids: ["geoId/06002", "geoId/06001"],
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: ["Count_Person"],
    });

  // get csv
  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "latest",
      minDate: "latest",
      parentPlace: "geoId/06",
      rowLimit: 7,
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });

  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "latest",
      minDate: "latest",
      parentPlace: "geoId/06",
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });

  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "",
      minDate: "2020",
      parentPlace: "geoId/06",
      rowLimit: 7,
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });

  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "",
      minDate: "2020",
      parentPlace: "geoId/06",
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });
}
