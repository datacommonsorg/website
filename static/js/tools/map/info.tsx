/**
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Info page before a chart is shown.
 */

import React, { useContext } from "react";

import { MemoizedInfoExamples } from "../shared/info_examples";
import { Context } from "./context";
import { ifShowChart } from "./util";

export function Info(): JSX.Element {
  const { statVar, placeInfo } = useContext(Context);

  return (
    <>
      {!ifShowChart(statVar.value, placeInfo.value) && (
        <div id="placeholder-container">
          <p>
            The map explorer helps you visualize how a statistical variable from
            the pane to the left can vary across geographic regions.
          </p>
          <ol>
            <li>
              Enter a place in the search box and then select the type of places
              you want to plot in the dropdown menu above.
            </li>
            <li>
              Pick a statistical variable in the left pane. There are thousands
              of statistical variables to choose from, arranged in a topical
              hierarchy.
            </li>
          </ol>
          <p>
            Or you can start your exploration from these interesting points ...
          </p>

          <MemoizedInfoExamples />

          <p>Take the data and use it on your site!</p>
          <p>
            <a href="mailto:collaborations@datacommons.org">Send</a> us your
            discoveries!
          </p>
        </div>
      )}
    </>
  );
}
