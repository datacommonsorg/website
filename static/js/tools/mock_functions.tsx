/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

jest.mock("axios");
import { when } from "jest-when";
import axios from "axios";
import hierarchy from "../../data/hierarchy_top.json";
import hierarchy_complete from "../../data/hierarchy_statsvar.json";
import { drawGroupLineChart } from "../chart/draw";
import * as d3 from "d3";

export function axios_mock(): void {
  // Mock all the async axios call.
  axios.get = jest.fn();

  // get hierachy.json
  when(axios.get)
    .calledWith("../../data/hierarchy_statsvar.json")
    .mockResolvedValue({ data: hierarchy });

  // get statsvar properties Median_Age_Person
  when(axios.get)
    .calledWith("/api/stats/stats-var-property?dcid=Median_Age_Person")
    .mockResolvedValue({
      data: {
        Median_Age_Person: { md: "", mprop: "age", pt: "Person", pvs: {} },
      },
    });

  // get statsVar info of Median_Age_Person and Count_Person
  when(axios.get)
    .calledWith(
      "/api/stats/stats-var-property?dcid=Median_Age_Person&dcid=Count_Person"
    )
    .mockResolvedValue({
      data: {
        Median_Age_Person: { md: "", mprop: "age", pt: "Person", pvs: {} },
        Count_Person: { md: "", mprop: "count", pt: "Person", pvs: {} },
      },
    });

  // get place stats vars, geoId/05
  when(axios.get)
    .calledWith("/api/place/statsvars/geoId/05")
    .mockResolvedValue({ data: ["Count_Person", "Median_Age_Person"] });

  // get place names, geoId/05
  when(axios.get)
    .calledWith("/api/place/name?dcid=geoId/05")
    .mockResolvedValue({ data: { "geoId/05": "Place" } });

  // get data, geoId/05, Median_Age_Person
  when(axios.get)
    .calledWith("/api/stats/Median_Age_Person?&dcid=geoId/05")
    .mockResolvedValue({
      data: {
        "geoId/05": {
          placeDcid: "geoId/05",
          placeName: "Arkansas",
          provenanceDomain: "census.gov",
          data: { "2011": 37.3, "2012": 37.4, "2013": 37.5, "2014": 37.6 },
        },
      },
    });

  // get data, geoId/05,Count_Person
  when(axios.get)
    .calledWith("/api/stats/Count_Person?&dcid=geoId/05")
    .mockResolvedValue({
      data: {
        "geoId/05": {
          placeDcid: "geoId/05",
          placeName: "Arkansas",
          provenanceDomain: "census.gov",
          data: { "1999": 37.3, "2010": 37.4, "2020": 37.5 },
        },
      },
    });
}

export function mock_hierarchy_complete(): void {
  axios.get = jest.fn();
  when(axios.get)
    .calledWith("../../data/hierarchy_statsvar.json")
    .mockResolvedValue({ data: hierarchy_complete });
}

// Mock drawGroupLineChart() as getComputedTextLength can has issue with jest
// and jsdom.
export function drawGroupLineChart_mock(): void {
  (drawGroupLineChart as jest.Mock).mockImplementation(
    (selector: string | HTMLDivElement) => {
      let container: d3.Selection<any, any, any, any>;
      if (typeof selector === "string") {
        container = d3.select("#" + selector);
      } else if (selector instanceof HTMLDivElement) {
        container = d3.select(selector);
      } else {
        return;
      }
      container.selectAll("svg").remove();

      const svg = container
        .append("svg")
        .attr("width", 500)
        .attr("height", 500);
      svg.append("text").text("svg text");
    }
  );
}
