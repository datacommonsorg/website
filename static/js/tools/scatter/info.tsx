/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Info page before a chart is shown.
 */

import React from "react";

import { MemoizedInfoExamples } from "../shared/info_examples";

function Info(): JSX.Element {
  return (
    <div id="placeholder-container">
      <p>
        The scatter plot tool helps you visualize the correlation between two
        statistical variables that appear in the pane to the left.
      </p>
      <ol>
        <li>
          Enter a place in the search box and then select the type of places you
          want to plot in the dropdown menu above.
        </li>
        <li>
          Pick two statistical variables in the left pane. There are thousands
          of statistical variables to choose from, arranged in a topical
          hierarchy.
        </li>
      </ol>
      <p>Or you can start your exploration from these interesting points ...</p>

      <MemoizedInfoExamples />

      <p>Take the data and use it on your site!</p>
      <p>
        <a href="mailto:collaborations@datacommons.org">Send</a> us your
        discoveries!
      </p>
    </div>
  );
}

export const MemoizedInfo = React.memo(Info);
