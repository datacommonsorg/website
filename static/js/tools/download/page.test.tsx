/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

jest.mock("axios");
jest.setTimeout(100000);

import { waitFor } from "@testing-library/react";
import axios from "axios";
import Enzyme, { mount, ReactWrapper } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import pretty from "pretty";
import React from "react";

import * as SharedUtil from "../../shared/util";
import { InfoPlace } from "./info";
import { axios_mock } from "./mock_functions";
import { Page } from "./page";

Enzyme.configure({ adapter: new Adapter() });

const globalAny: any = global;

const INFO_PLACES: [InfoPlace, InfoPlace] = [
  { name: "Place 1", dcid: "dcid/1" },
  { name: "Place 2", dcid: "dcid/2" },
];

async function waitForComponentUpdates(wrapper: ReactWrapper) {
  // Wait for state updates
  await waitFor(() => {
    expect(wrapper.text()).toContain("");
  });
  // Wait for stat var info and place info fetching
  await new Promise(process.nextTick);
  wrapper.update();
  await new Promise(process.nextTick);
  wrapper.update();
}

test("Loading options from URL", async () => {
  globalAny.window = Object.create(window);
  // Set url hash
  Object.defineProperty(window, "location", {
    writable: true,
    value: {
      hash: "#place=geoId/06&pt=County&sv=Count_Person",
    },
  });
  // Mock all the async axios calls
  axios_mock();
  // Render the component
  const wrapper = mount(<Page infoPlaces={INFO_PLACES} />);
  await waitForComponentUpdates(wrapper);
  // Check that preview table shows up on click and matches snapshot
  wrapper.find(".get-data-button").at(0).simulate("click");
  await waitForComponentUpdates(wrapper);
  await waitFor(() => {
    expect(wrapper.text()).toContain("placeDcid");
  });
  expect(
    pretty(wrapper.find("#preview-section").at(0).getDOMNode().innerHTML)
  ).toMatchSnapshot();
  // Check that clicking download gets the right data and calls the saveToFile function.
  wrapper.find(".download-button").at(0).simulate("click");
  expect(axios.post).toHaveBeenCalledWith("/api/csv/within", {
    childType: "County",
    facetMap: {
      Count_Person: "",
    },
    maxDate: "latest",
    minDate: "latest",
    parentPlace: "geoId/06",
    statVars: ["Count_Person"],
  });
  let savedFile = "";
  jest.spyOn(SharedUtil, "saveToFile").mockImplementation((fileName) => {
    savedFile = fileName;
  });
  await waitFor(() => {
    expect(savedFile).toEqual("California_County.csv");
  });
});

test("Manually updating options", async () => {
  globalAny.window = Object.create(window);
  // Set url hash
  Object.defineProperty(window, "location", {
    writable: true,
    value: {
      hash: "#place=geoId/06",
    },
  });
  // Mock all the async axios calls
  axios_mock();
  // Render the component
  const wrapper = mount(<Page infoPlaces={INFO_PLACES} />);
  await waitForComponentUpdates(wrapper);
  // Choose place type
  wrapper
    .find("#place-selector-place-type")
    .at(0)
    .simulate("change", { target: { value: "County" } });
  await waitFor(() => {
    expect(axios.get).toHaveBeenCalledWith(
      "/api/place/places-in?dcid=geoId/06&placeType=County"
    );
  });
  await waitForComponentUpdates(wrapper);
  // Update Date Range
  wrapper.find("#date-range").at(0).simulate("click");
  wrapper.update();
  wrapper
    .find(".download-date-range-input")
    .at(0)
    .simulate("change", { target: { value: "2020" } });
  wrapper.update();
  // Add stat vars using stat var widget
  wrapper
    .find("#hierarchy-section .Collapsible__trigger")
    .at(0)
    .simulate("click");
  await waitForComponentUpdates(wrapper);
  wrapper
    .find("#hierarchy-section input")
    .at(0)
    .simulate("change", { target: { checked: true } });
  await waitForComponentUpdates(wrapper);
  wrapper
    .find("#hierarchy-section input")
    .at(1)
    .simulate("change", { target: { checked: true } });
  await waitForComponentUpdates(wrapper);
  // Delete one stat var using the chips
  wrapper.find(".download-sv-chips .chip-action i").at(1).simulate("click");
  await waitForComponentUpdates(wrapper);
  // Check that preview table shows up on click and matches snapshot
  wrapper.find(".get-data-button").at(0).simulate("click");
  await waitForComponentUpdates(wrapper);
  await waitFor(() => {
    expect(wrapper.text()).toContain("placeDcid");
  });
  expect(
    pretty(wrapper.find("#preview-section").at(0).getDOMNode().innerHTML)
  ).toMatchSnapshot();
  // Check that clicking download gets the right data and calls the saveToFile function.
  wrapper.find(".download-button").at(0).simulate("click");
  expect(axios.post).toHaveBeenCalledWith("/api/csv/within", {
    childType: "County",
    facetMap: {
      Count_Person: "",
    },
    maxDate: "",
    minDate: "2020",
    parentPlace: "geoId/06",
    statVars: ["Count_Person"],
  });
  let savedFile = "";
  jest.spyOn(SharedUtil, "saveToFile").mockImplementation((fileName) => {
    savedFile = fileName;
  });
  await waitFor(() => {
    expect(savedFile).toEqual("California_County.csv");
  });
});
