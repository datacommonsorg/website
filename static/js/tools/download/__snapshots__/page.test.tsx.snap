// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading options from URL 1`] = `
"<table>
  <thead>
    <tr>
      <th>placeDcid</th>
      <th>placeName</th>
      <th>Date:Count_Person</th>
      <th>Value:Count_Person</th>
      <th>Source:Count_Person</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>geoId/06001</td>
      <td>Alameda County</td>
      <td>2021</td>
      <td>1648556</td>
      <td>https://www2.census.go</td>
    </tr>
    <tr>
      <td>geoId/06002</td>
      <td>Alpine County</td>
      <td>2021</td>
      <td>1235</td>
      <td>https://www2.census.gov/</td>
    </tr>
    <tr>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
  </tbody>
</table><button type=\\"button\\" class=\\"download-button btn btn-primary\\">Download</button>
<div class=\\"screen\\">
  <div id=\\"spinner\\"></div>
</div>"
`;

exports[`Manually updating options 1`] = `
"<table>
  <thead>
    <tr>
      <th>placeDcid</th>
      <th>placeName</th>
      <th>Date:Count_Person</th>
      <th>Value:Count_Person</th>
      <th>Source:Count_Person</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>geoId/06001</td>
      <td>Alameda County</td>
      <td>2021</td>
      <td>1648556</td>
      <td>https://www2.census.go</td>
    </tr>
    <tr>
      <td>geoId/06002</td>
      <td>Alpine County</td>
      <td>2021</td>
      <td>1235</td>
      <td>https://www2.census.gov/</td>
    </tr>
    <tr>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
  </tbody>
</table><button type=\\"button\\" class=\\"download-button btn btn-primary\\">Download</button>
<div class=\\"screen\\">
  <div id=\\"spinner\\"></div>
</div>"
`;
